src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs

```diff
@@ -0,0 +1,351 @@
+            }
+
+            Assert.Equal(input, destination.ToArray());
+            Assert.Equal(input.Length, destination.Position);
+            Assert.Equal(input.Length, destination.Length);
+        }
+
+        [Theory]
+        [MemberData(nameof(TwoBooleans))]
+        public async Task NonSeekableMemoryStream_To_MemoryStream(bool sourceIsExposable, bool disposeSource)
+        {
+            byte[] input = CreateByteArray(8192);
+            var source = new NonSeekableMemoryStream(input, sourceIsExposable);
+            var destination = new MemoryStream();
+
+            await StreamToStreamCopy.CopyAsync(source, destination, 4096, disposeSource);
+
+            Assert.NotEqual(disposeSource, source.CanRead);
+            if (!disposeSource)
+            {
+                Assert.Equal(input.Length, source.Position);
+            }
+
+            Assert.Equal(input, destination.ToArray());
+            Assert.Equal(input.Length, destination.Position);
+            Assert.Equal(input.Length, destination.Length);
+        }
+
+        [Theory]
+        [MemberData(nameof(TwoBooleans))]
+        public async Task MemoryStream_NonZeroPosition_To_MemoryStream(bool sourceIsExposable, bool disposeSource)
+        {
+            byte[] input = CreateByteArray(8192);
+            MemoryStream source = CreateSourceMemoryStream(sourceIsExposable, input);
+            const int StartingPosition = 1024;
+            source.Position = StartingPosition;
+
+            var destination = new HttpContent.LimitMemoryStream(int.MaxValue);
+
+            await StreamToStreamCopy.CopyAsync(source, destination, 0x1000, disposeSource);
+
+            Assert.NotEqual(disposeSource, source.CanRead);
+            if (!disposeSource)
+            {
+                Assert.Equal(input.Length, source.Position);
+            }
+
+            Assert.Equal(input.Skip(StartingPosition), destination.ToArray());
+            Assert.Equal(input.Length - StartingPosition, destination.Position);
+            Assert.Equal(input.Length - StartingPosition, destination.Length);
+        }
+
+        [Theory]
+        [MemberData(nameof(TwoBooleans))]
+        public async Task MemoryStream_PositionAtEnd_To_MemoryStream(bool sourceIsExposable, bool disposeSource)
+        {
+            byte[] input = CreateByteArray(8192);
+            MemoryStream source = CreateSourceMemoryStream(sourceIsExposable, input);
+            int StartingPosition = input.Length;
+            source.Position = StartingPosition;
+
+            var destination = new HttpContent.LimitMemoryStream(int.MaxValue);
+
+            await StreamToStreamCopy.CopyAsync(source, destination, 0x1000, disposeSource);
+
+            Assert.NotEqual(disposeSource, source.CanRead);
+            if (!disposeSource)
+            {
+                Assert.Equal(input.Length, source.Position);
+            }
+
+            Assert.Equal(input.Skip(StartingPosition), destination.ToArray());
```

Comment:
> Be careful, Steve, you'll scold you for not using Array.Empty<byte>() here

Where?

> All of these tests look rather similar from (e.g.) 94-102; perhaps that should be a utility assertion method that you defer to after doing the setup?

There are subtle differences in each, and the helper would likely need to take in most of the values being passed to each assert.  Didn't seem worthwhile.